Rework the viewport animation functions

starwed opened the pull request:
This isn't quite ready to land yet, but thoguht I'd open the PR so folk could see the changes.  This alters several aspects of the viewport:
- Changes pan and zoom to work with Crafty.easing. (Time based instead of tick based.)
- Makes sure the zoom animation is nice and smooth.  (Surprisingly tricky)
- Fixes a bug in how scale is treated by the DOM/Canvas renderers
- Adds a new "StopCamera" event for stopping any extant camera animation (and uses it when a new animation is requested.)
- Fires an "CameraAnimationDone" event when a camera animation completes

Along the way this fixes #163, #285, #339, and #616

starwed commented:
Ok, this is a candidate to land now.  I had to fix some issues with the tests.

Forgot to mention that this changes the way you use `pan()`, by specifying a direction along both axes at once, rather than specifying an axis individually.

There was a feature that forced the viewport to stick to integer coordinates, to prevent weird aliasing issues.  But this was actually really ugly when zooming slowly; it caused the position to stutter a lot.  So I disable this, but we should probably allow it as optional behavior.

kevinsimper commented on src/drawing.js:
```diff
@@ -520,7 +520,7 @@ Crafty.DrawManager = (function () {
             if (dirtyViewport) {
                 var view = Crafty.viewport;
-                ctx.setTransform(view._scale, 0, 0, view._scale, view.x, view.y);
+                ctx.setTransform(view._scale, 0, 0, view._scale, view._x*view._scale, view._y*view._scale);
```
This solves the bug where the zoom is going southeast?
http://www.w3schools.com/tags/canvas_settransform.asp

starwed replied:
Probably, but since I rewrote the zoom code at the same time I'm not 100% sure it was the cause of the original bug.

kevinsimper commented:
:+1:  Nice

But this pull request could indeed need different commit because it can be a little hard which changes are solving what problems :-)

Like the code together with the test that would be awesome, some of this do not make sense unless your are really dig into canvas and use time reading the code. 

There is a reason not everybody is doing this (because it is friggin' hard, and canvas is still "new" so every changes for learning something new is appreciated! :+1:

starwed commented:
Heh, I just fixed things until the zoom code worked, which is why it's all one lump.

The change to the drawing code could definitely be a separate commit.

starwed committed:
Fix how drawing code handles scale

The existing draw code doesn't take into account that the scaling happens before translation.  This patch fixes that by multiplying the translation by the current scale.

starwed committed:
Rework several viewport animation functions

- Changes pan and zoom to work with Crafty.easing. (Time based instead of tick based.)
- Makes sure the zoom animation is nice and smooth.  (Surprisingly tricky)
- Adds a new "StopCamera" event for stopping any extant camera animation (and uses it when a new animation is requested.)
- Fires an "CameraAnimationDone" event when a camera animation completes

Fixes #163, #285, #339, and #616

starwed commented:
Ok, split the drawing commit out, and added some tests for the CameraAnimationDone event.

starwed commented:
Going to go ahead and merge this.
